### DISCUSSION TITLE
Improve options when context length is exceeded

### DISCUSSION THREAD

**Discussion by heyitstonyg** [cline/cline#499] [Feature Requests]
Currently when you exceed the maximum context length you get a message like the one below and only 2 options: "Retry" or "Start new task". And the text entry field is disabled.

Could there be a 3rd option? Or could the text entry field be enabled so we can type another message

If I "Retry" it will send the same message and fail again, but if I start new task then I need to start the task over again. 

> 
"This endpoint's maximum context length is 128000 tokens. However, you requested about 419745 tokens (419745 of text input). Please reduce the length of either one, or use the "middle-out" transform to compress your prompt automatically."

**Comment by stuinfla** [comment]
I agree this is a huge issue for me. I have sample files that I'm having it load and inadvertently load one and then blow out of the whole task before I've had it had a chance to update the history in the context and then it's lost and I'm starting over and I've lost all the time doing that task and I'm not sure what state I'm in and it's a real mess. We really need a cleaner solution for dealing with this.. Preferably one that figures out that reading a file would be too big and it would blow out the context. This has happened several times like when I've tried to load large pictures of screenshots to show what was going on on a browser or also loading large 300 K Chase on files for make.com scenarios.

**Comment by vietvudanh** [comment]
Same scenario for me, I had some data files like csv that I need to write python code to process data. Natural ways of course is too add the files and ask Cline to do it.
However since the files is too big, context failed. The workaround I am doing is to head some lines of the original files to a sample file and work on that, but it would be nice if can provide options for Cline to read just some data lines.

**Reply by kowkowhuang** [reply]
Same here. This is a very good feature Same scenario for me, I had some data files like csv that I need to write python code to process data. Natural ways of course is too add the files and ask Cline to do it.
However since the files is too big, context failed. The workaround I am doing is to head some lines of the original files to a sample file and work on that, but it would be nice if can provide options for Cline to read just some data lines.

**Reply by vietvudanh** [reply]
I had tried windsurf, their agent default action is head some data liens (5), works really well.

**Comment by lloydchang** [comment]
relates to https://discord.com/channels/1275535550845292637/1275535550845292640/1311442009403162665
and https://discord.com/channels/1275535550845292637/1275535550845292640/1311459959162146886
> https://openrouter.ai/docs/transforms for openrouter's real thing, _humorously named_ "`middle-out`" by combining hbo's silicon valley and https://arxiv.org/abs/2307.03172 according to https://x.com/xanderatallah/status/1678511019834896386

plus https://discord.com/channels/1275535550845292637/1275535550845292640/1311469773208420492
> if anyone wants to experiment with https://openrouter.ai/docs/transforms 
```TypeScript
transforms: ["middle-out"], // Compress prompts > context size. This is the default for endpoints with context length <= 8k
```
you would modify https://github.com/cline/cline/blob/main/src/api/providers/openrouter.ts

**Comment by geezz07** [comment]
When rate limit is almost reached can we have Cline give us an option to summarize what has been done to that point and start a new window with the summary so it can continue the task?

**Comment by pauladkisson** [comment]
Adding an option for middle-out would be super helpful for tasks where I need to make small edits to lots of files. Alternatively, cline could split up a task into subtasks and use a separate LLM instance for each subtask?

**Comment by plymouthvan** [comment]
This issue is a significant headache. Especially if a task is allowed to run with auto approve when 99% of the task can run on its own, but if Cline for whatever reason wants to try and read a file that is too big, it will seize up. If checkpoints have been created, then you can return to the last checkpoint, but no checkpoints are created in Plan mode, and so if it spends 10 minutes on a plan and then seizes up, you're starting over completely. Not to mention Cline's planning process can rack up API costs while it's figuring out what to do and then if it exceeds its prompt length, that money is just completely wasted. 

- The simplest solution really would be to just provide a third option to reply and forget the last prompt Cline tried to send. Most of the time all a user would have to say is, "Sorry you can't look directly at such and such file" and Cline would just move on.

- Another solution would be to create chat restoration checkpoints at every step Cline takes. As it is, checkpoints are only created when Cline edits a file.

This feels like it should be a fairly high priority, particularly considering the solutions seem rather easy to implement.

**Comment by CarlosCommits** [comment]
This. I think it would be nice to have the option to delete messages/specific context. Often times Cline will try to open/read a large .md log file and break it's context. Then I can't remove it from the chat so I have to start a new one

**Comment by skuta** [comment]
can cline FORECAST this would happen and stop you in sending another message and think twice what to ask from the LLM next?

**Comment by jellymellio** [comment]
A built-in memory bank or a summarize function, where a new API call with a large context window summarizes everything done and saves it, so you can start fresh without loosing context would solve this.

**Comment by mario-guerra** [comment]
Chiming in to say I'm seeing this issue as well, it would be awesome to have an option to remove old messages from the history or summarize the discussion in order to continue.

**Comment by afinetooth** [comment]
This may have been addressed as of Apr 15, 2025 with this CLINE blog post:

- [Unlocking Persistent Memory: How Cline's new_task Tool Eliminates Context Window Limitations](https://cline.bot/blog/unlocking-persistent-memory-how-clines-new_task-tool-eliminates-context-window-limitations)

And this doc on the `/new_task` tool:

- [New Task Tool](https://docs.cline.bot/exploring-clines-tools/new-task-tool)

**Comment by santijpg** [comment]
You can use Memory Bank [Memory Bank Cline](https://docs.cline.bot/prompting/cline-memory-bank)

_The Memory Bank is a structured documentation system that allows Cline to maintain context across sessions. It transforms Cline from a stateless assistant into a persistent development partner that can effectively “remember” your project details over time._

**Comment by ferhimedamine** [comment]
This might be an unpopular take, but I think the deeper fix isn't a better UI for "context exceeded" — it's preventing the context from overflowing in the first place by giving the agent external memory. The agent shouldn't need to hold everything in the context window simultaneously.

The pattern that's worked for us: store facts, code patterns, and decisions to a persistent backend as the session progresses, then retrieve only what's relevant for the current step via semantic search + BM25 hybrid. The context window becomes a working-memory scratchpad rather than a complete history log. This also naturally handles multi-session workflows since the memory persists across restarts.

We open-sourced the self-hosted setup for this if anyone wants to experiment: https://github.com/Dakera-AI/dakera-deploy
